src/asmcnc/comms/serial_connection.py

```diff
@@ -519,60 +517,96 @@ def _reset_counters(self):
     def set_streaming_flags_to_true(self):
         # self.m.set_pause(False) # moved to go screen for timing reasons
         self.is_stream_lines_remaining = True
-        self.is_job_streaming = True    # allow grbl_scanner() to start stuffing buffer
+        self.is_job_streaming = True  # allow grbl_scanner() to start stuffing buffer
         log('Job running')
-    
-    def stuff_buffer(self): # attempt to fill GRBLS's serial buffer, if there's room      
+
+    def stuff_buffer(self):  # attempt to fill GRBLS's serial buffer, if there's room
 
         while self.l_count < len(self.jd.job_gcode_running):
-            
+
             line_to_go = self.add_line_number_to_gcode_line(self.jd.job_gcode_running[self.l_count], self.l_count)
             serial_space = self.RX_BUFFER_SIZE - sum(self.c_line)
-    
+
             # if there's room in the serial buffer, send the line
             if len(line_to_go) + 1 <= serial_space:
-                self.c_line.append(len(line_to_go) + 1) # Track number of characters in grbl serial read buffer
-                self.write_direct(line_to_go, show_in_sys = True, show_in_console = False) # Send g-code block to grbl
-                self.l_count += 1 # lines sent to grbl           
+                self.scrape_last_sent_modes(line_to_go)
+                self.add_to_g_mode_tracker(self.last_sent_motion_mode, self.last_sent_feed, self.last_sent_speed)
+                self.c_line.append(len(line_to_go) + 1)  # Track number of characters in grbl serial read buffer
+                self.write_direct(line_to_go, show_in_sys=True, show_in_console=False)  # Send g-code block to grbl
+                self.l_count += 1  # lines sent to grbl
             else:
                 return
- 
+
         self.is_stream_lines_remaining = False
 
+    ## GCODE BUFFER STUFFING TRACKING
+    last_line_executed = 0
+
+    last_sent_motion_mode = ""
+    last_sent_feed = 0
+    last_sent_speed = 0
+
+    feed_pattern = re.compile(r"F\d+\.?\d*")
+    speed_pattern = re.compile(r"S\d+\.?\d*")
+    g_motion_pattern = re.compile(r"G([0-3])(\D|$)")
```

Comment:
I think gcodes like G00 and G01 are valid, so maybe change to G0?([0-3])(\D|$)

Reply:
G0?([0-3])(\D|$) doesn't seem to work correctly (it's fine when I test it on https://regex101.com/ but not with python 2.7, so I suspect that there's some difference there), as it returns the "0" as the first group. Investigating now...